language/move-stdlib/nursery/sources/Capability.move

```diff
@@ -0,0 +1,176 @@
+/// predicate:
+///
+/// ```
+///   invariant forall a: address where exists<CapState<ProtectedFeature>>(addr):
+///               forall d in Capability::spec_delegates<ProtectedFeature>(a):
+///                  is_valid_delegate_for_protected_feature(d);
+/// ```
+///
+module Std::Capability {
+    use Std::Errors;
+    use Std::Signer;
+    use Std::Vector;
+
+    const ECAP: u64 = 0;
+    const EDELEGATE: u64 = 1;
+
+    /// The token representing an acquired capability. Cannot be stored in memory, but copied and dropped freely.
+    struct Cap<phantom Feature> has copy, drop {
+        root: address
+    }
+
+    /// An internal data structure for representing a configured capability.
+    struct CapState<phantom Feature> has key {
+        delegates: vector<address>
+    }
+
+    /// An internal data structure for representing a configured delegated capability.
+    struct CapDelegateState<phantom Feature> has key {
+        root: address
+    }
+
+    /// Creates a new capability class, owned by the passed signer.
+    public fun create<Feature>(owner: &signer) {
+        let addr = Signer::address_of(owner);
+        assert(!exists<CapState<Feature>>(addr), Errors::already_published(ECAP));
+        move_to<CapState<Feature>>(owner, CapState{ delegates: Vector::empty() });
+    }
+
+    /// Acquires a capability token. Only the owner of the capability class, or an authorized delegate,
+    /// can succeed with this operation.
+    public fun acquire<Feature>(requester: &signer): Cap<Feature>
+    acquires CapState, CapDelegateState {
+        let addr = Signer::address_of(requester);
+        if (exists<CapDelegateState<Feature>>(addr)) {
+            let root_addr = borrow_global<CapDelegateState<Feature>>(addr).root;
+            // double check that requester is actually registered as a delegate
+            assert(exists<CapState<Feature>>(root_addr), Errors::invalid_state(EDELEGATE));
+            assert(Vector::contains(&borrow_global<CapState<Feature>>(root_addr).delegates, &addr),
+                   Errors::invalid_state(EDELEGATE));
+            Cap<Feature>{root: root_addr}
+        } else {
+            assert(exists<CapState<Feature>>(addr), Errors::not_published(ECAP));
+            Cap<Feature>{root: addr}
+        }
+    }
+
+    /// Registers a delegation relation.
+    public fun delegate<Feature>(cap: Cap<Feature>, to: &signer)
+    acquires CapState {
+        let addr = Signer::address_of(to);
+        assert(!exists<CapDelegateState<Feature>>(addr), Errors::already_published(EDELEGATE));
+        assert(exists<CapState<Feature>>(cap.root), Errors::invalid_state(ECAP));
+        move_to(to, CapDelegateState<Feature>{root: cap.root});
+        add_element(&mut borrow_global_mut<CapState<Feature>>(cap.root).delegates, addr);
+    }
+
+    /// Revokes a delegation relation.
+    public fun revoke<Feature>(cap: Cap<Feature>, from: address)
```

Comment:
I think we may also need a way for an account to remove a delegated capability from its account even after a revocation has occured. Otherwise the account is stuck with that capability until the granting account decides to remove it.

Reply:
In fact the delegate can self-revoke. 

```
fun revoke_me(s: &signer) {
   let cap = Capability::acquire<T>(s);
   Capability::revoke(cap, Signer::address_of(s));
}
```

Currently, everybody who can acquire the cap can revoke it for everybody, not only the original owner.

Reply:
Ah good to know! I guess my worry then is if the account has a revoked capability they have no way of removing it from under their account (at least as far as I can see). In particular in the example you gave:

```rust
fun revoke_me(s: &signer) {
   let cap = Capability::acquire<T>(s); // This line with abort due to the check at line 128
   Capability::revoke(cap, Signer::address_of(s));
}
```

Reply:
I tried at least to design this so revocation will always remove both the resource at the delegate as well as the entry in the list. So there shouldn't be a state possible where the capability is revoked _and_ the resource still exists (modulo bugs).

It might be though interesting whether we should make revoke idempotent.